### ISSUE TITLE
openAI generates samples^2, instead of samples when using `generate_base`.

### ISSUE THREAD

**Issue by HerrIvan** [outlines-dev/outlines#363]
### Describe the issue as clearly as possible:

I paste below code from `outlines/models/openai.py`, function `generate_base`. In my opinion it contains a bug: it generates `samples*samples` per iteration , instead of simply `samples`. (See comments).

```python
            decoded_samples = []
            for _ in range(samples):    # <--- We are iterating over samples
                decoded: List[str] = []
                for i in range(max([len(word) for word in encoded])):
                    
                   [...]

                    response = await call_api(
                        model_name,
                        format_prompt(prompt),
                        1,
                        temperature,
                        [],
                        mask,
                        samples,  # <---- But we are querying again for n samples again
                        api_key,
                    )
                    decoded.append(extract_choice(response["choices"][0])). # <--- But only storing the first
                    prompt = prompt + "".join(decoded)

                decoded_samples.append("".join(decoded))
```

The fix would simply be to replace `samples` by `1` in the arguments of the function call.

### Steps/code to reproduce the bug:

```python
Not necessary to run the code.
```

### Expected result:

```shell
The result, in terms of output will be the same, but less tokens will be consumed.
```

### Outlines/Python version information:

```python
Outlines: 0.0.10
```

### Context for the issue:

I would have written a pull request, but it seems an overkill for replacing one character in one line of code. I can still do it if that reduces your work.

**Comment by rlouf** [comment]
Thank you for the report! For performance purposes we should remove the loop `for _ in range(samples)` and instead ask the API to return several samples. It's a slightly more involved change, but should be straightforward nevertheless. I would greatly appreciate it if you could open a RP :)

**Comment by HerrIvan** [comment]
You're welcome.

But does it really work that way in parallel? 

Say I first calls the API with `n=samples`, I may get `n` different first tokens. But after that I cannot query for `n` samples again, as I would then be creating `samples`^`max_tokens` different sequences at the end of the generation.

Do you have a similar pattern implemented somewhere else in the code? It must be a general issue for the interaction of guided generation one token at a time + multiple samples. If so, if you point me to it and I could borrow from that.

**Comment by HerrIvan** [comment]
Closed with #376
